pkg/ruler/notifier.go

```diff
@@ -27,18 +28,43 @@ import (
 	"github.com/grafana/mimir/pkg/util"
 )
 
-var errRulerNotifierStopped = cancellation.NewErrorf("rulerNotifier stopped")
+var (
+	errRulerNotifierStopped               = cancellation.NewErrorf("rulerNotifier stopped")
+	errRulerSimultaneousBasicAuthAndOAuth = errors.New("cannot use both Basic Auth and OAuth2 simultaneously")
+)
 
 type NotifierConfig struct {
 	TLSEnabled bool             `yaml:"tls_enabled" category:"advanced"`
 	TLS        tls.ClientConfig `yaml:",inline"`
 	BasicAuth  util.BasicAuth   `yaml:",inline"`
+	OAuth2     OAuth2Config     `yaml:"oauth2"`
+	ProxyURL   string           `yaml:"proxy_url" category:"advanced"`
 }
 
 func (cfg *NotifierConfig) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.TLSEnabled, "ruler.alertmanager-client.tls-enabled", true, "Enable TLS for gRPC client connecting to alertmanager.")
 	cfg.TLS.RegisterFlagsWithPrefix("ruler.alertmanager-client", f)
 	cfg.BasicAuth.RegisterFlagsWithPrefix("ruler.alertmanager-client.", f)
+	cfg.OAuth2.RegisterFlagsWithPrefix("ruler.alertmanager-client.oauth.", f)
+	f.StringVar(&cfg.ProxyURL, "ruler.alertmanager-client.proxy-url", "", "Optional HTTP, HTTPS via CONNECT, or SOCKS5 proxy URL to route requests through. Applies to all requests, including auxiliary traffic, such as OAuth token requests.")
+}
+
+type OAuth2Config struct {
+	ClientID     string              `yaml:"client_id"`
+	ClientSecret flagext.Secret      `yaml:"client_secret"`
+	TokenURL     string              `yaml:"token_url"`
+	Scopes       flagext.StringSlice `yaml:"scopes,omitempty"`
```

Comment:
Note: Moving this to StringSliceCSV. Nearly every other field that behaves this way across the project uses the CSV version, the regular version looks like it doesn't work as cleanly.